lib/spack/spack/spec.py

```diff
@@ -2095,31 +2127,109 @@ def traverse_edges(
-            current_name = item.spec.name
-            new_name = spack.aliases.BUILTIN_TO_LEGACY_COMPILER.get(current_name, current_name)
-            # note: depflag not allowed, currently, on "direct" edges
-            edge_attributes = ""
-            if item.virtuals or item.when != Spec():
-                edge_attributes = item.spec.format("{edge_attributes}") + " "
-
-            parts.append(f"%{edge_attributes}{item.spec.format()}".replace(current_name, new_name))
-        for item in sorted(transitive, key=lambda x: x.spec.name):
-            # Recurse to attach build deps in order
-            edge_attributes = ""
-            if item.virtuals or item.depflag or item.when != Spec():
-                edge_attributes = item.spec.format("{edge_attributes}") + " "
-            parts.append(f"^{edge_attributes}{str(item.spec)}")
+
+        # helper for direct and transitive loops below
+        def format_edge(edge, sigil, dep_spec=None):
+            dep_spec = dep_spec or edge.spec
+            dep_format = dep_spec.format(format_string)
+
+            edge_attributes = (
+                self._format_edge_attributes(edge, deptypes=deptypes, virtuals=False)
+                if edge.depflag or edge.when != Spec()
+                else ""
+            )
+            virtuals = f"{','.join(edge.virtuals)}=" if edge.virtuals else ""
+            star = _anonymous_star(edge, dep_format)
+
+            return f"{sigil}{edge_attributes}{star}{virtuals}{dep_format}"
+
+        # direct dependencies
+        for edge in sorted(direct, key=lambda x: x.spec.name):
+            if not include(edge):
+                continue
+
+            # replace legacy compiler names
+            old_name = edge.spec.name
+            new_name = spack.aliases.BUILTIN_TO_LEGACY_COMPILER.get(old_name)
+            try:
+                # this is ugly but copies can be expensive
+                if new_name:
+                    edge.spec.name = new_name
+                parts.append(format_edge(edge, "%", edge.spec))
+            finally:
+                edge.spec.name = old_name
+
+        # transitive dependencies (with any direct dependencies)
+        for edge in sorted(transitive, key=lambda x: x.spec.name):
+            if not include(edge):
+                continue
+            sigil = "%" if _force_direct else "^"  # hack til direct deps represented better
+            parts.append(format_edge(edge, sigil, edge.spec))
+
+            # also recursively add any direct dependencies of transitive dependencies
+            if edge.spec._dependencies:
+                parts.append(
+                    edge.spec._format_dependencies(
+                        format_string=format_string,
+                        include=include,
+                        deptypes=deptypes,
+                        _force_direct=_force_direct,
+                    )
+                )
+
         return " ".join(parts).strip()
 
+    @property
+    def compilers(self):
```

Comment:
Why calling it `compilers`, given we already have `Spec.compiler`, and not making this function name start with an underscore?